dashboard/modules/job/tests/test_sdk.py

```diff
@@ -183,18 +182,56 @@ def get_register_agents_number():
     def submit_job_and_wait_finish():
         submission_id = client.submit_job(entrypoint="echo hello")
 
-        wait_for_condition(_check_job_succeeded, client=client, job_id=submission_id)
+        wait_for_condition(
+            _check_job_succeeded, client=client, job_id=submission_id, timeout=30
+        )
 
-    # make sure list(cluster.worker_nodes)[0] will be the owner of a supervisor actor.
-    cluster.add_node(dashboard_agent_listen_port=52366)
+    worker_1_http_port = 52366
+    cluster.add_node(dashboard_agent_listen_port=worker_1_http_port)
     wait_for_condition(lambda: get_register_agents_number() == 2, timeout=20)
     assert len(cluster.worker_nodes) == 1
     node_try_to_kill = list(cluster.worker_nodes)[0]
-    submit_job_and_wait_finish()
 
-    cluster.add_node(dashboard_agent_listen_port=52367)
+    def make_sure_worker_node_run_job(port):
+        actors = ray.state.actors()
+
+        def _kill_all_driver():
+            for _, actor_info in actors.items():
+                if actor_info["State"] != "ALIVE":
+                    continue
+                if actor_info["Name"].startswith("_ray_internal_job_actor"):
+                    proc = psutil.Process(actor_info["Pid"])
+                    try:
+                        proc.kill()
+                    except Exception:
+                        pass
+
+        try:
+            for _, actor_info in actors.items():
+                if actor_info["State"] != "ALIVE":
+                    continue
+                if actor_info["Name"].startswith("_ray_internal_job_actor"):
+                    proc = psutil.Process(actor_info["Pid"])
+                    parent_proc = proc.parent()
+                    if f"--listen-port={port}" in " ".join(parent_proc.cmdline()):
+                        _kill_all_driver()
+                        return True
+        except Exception as ex:
+            print("Got exception:", ex)
+            raise
+        client.submit_job(entrypoint="sleep 3600")
+        return False
+
+    # Make list(cluster.worker_nodes)[0] called at least once
+    wait_for_condition(
+        lambda: make_sure_worker_node_run_job(worker_1_http_port), timeout=60
+    )
+
+    worker_2_http_port = 52367
+    cluster.add_node(dashboard_agent_listen_port=worker_2_http_port)
     wait_for_condition(lambda: get_register_agents_number() == 3, timeout=20)
 
+    # The third `JobAgent` will not be called here.
```

Comment:
I didn't understand this part, why won't the third jobagent be called here?  Couldn't the third jobagent be the result of the random agent choice?

Reply:
@architkulkarni What I wrote here is not very rigorous. In short, CANDIDATE_AGENT_NUMBER is 2, the agent on the head node and the agent on the first worker have been selected, so the agent of the second worker, that is, the third agent not will be used

Reply:
I push a new commit to make sure the agent on the head node has been selected: [`1e05ac7` (#29008)](https://github.com/ray-project/ray/pull/29008/commits/1e05ac7c4c8e9a11aa6d67d7047db75afe51dbb1)